sensu/plugins/check-swift-dispersion.py

```diff
@@ -35,22 +35,37 @@ def setup(self):
                                  help='Object warning dispersion percent')
 
     def run(self):
-
         output = None
-        try:
-            output = check_output(['swift-dispersion-report', '-j'],
-                                  stderr=STDOUT)
-
-            # strip any ERRORs that come through
-            p = re.compile(r'(\{.*\})')
-            m = p.search(output)
-            output = m.group(1)
-
-        except CalledProcessError as e:
-            self.critical("Unable to run swift-dispersion-check: %s %s" %
-                          (e, output))
-            return
-
+        # This should at most run twice when populate is not
+        # run. If populate is run error message would be different
+        # causing check to return immediately
+        while True:
+            proc_rep = Popen(['swift-dispersion-report', '-j'],
+                             stdout=PIPE, stderr=PIPE)
+            output, error = proc_rep.communicate()
+            if proc_rep.returncode == 0:
+                break
+            else:
+                # If dispersion populate is not run report returns error saying
+                # no containers present. So once populate is run successfully
+                # this should error out if report fails
+                if "No containers to query" not in error:
+                    self.critical("Unable to run swift-dispersion-check: %s" %
+                                  error)
+                    return
+                else:
+                    # Run dispersion populate and retry
+                    proc_pop = Popen(['swift-dispersion-populate'],
+                                     stdout=PIPE, stderr=PIPE)
+                    pop_out, pop_error = proc_pop.communicate()
+                    # Error out if populate fails for some reason
+                    if proc_pop.returncode != 0:
+                        self.critical("Unable to run swift-dispersion-check: "
```

Comment:
never try to fix a problem inside a sensu check ...  instead use the remediation handler ...  make auto-remediation a feature flag so that users can disable it via ansible variable.